feat(overrides): prop overrides accept a function too

qqingvarqq opened the pull request:
Resolves #1552 <!-- remove the (`) quotes to link the issues -->

#### Description
Make it possible to pass a function into props overrides. 
Rename `mergeStyleOverrides` to `mergeSpecialOverrides` since the same logic apply to `styles` and `props` overrides.
One important thing, since `mergeStyleOverrides` is only used inside `overrides.js` file I suggest that it is an internal function, so renaming it, is not breaking changes. (If it is breaking changes, we can define an alias for `mergeSpecialOverrides` as  `mergeStyleOverrides` and remove him when the new major version released)
Also, updated flow types, updated unit tests for `getOverrideProps` and  `getOverrides`.
<!-- Describe your changes below in as much detail as possible -->

#### Scope
- [X] Minor: New Feature
- [X] Major: Maybe Breaking Change

chasestarr commented on src/helpers/__tests__/overrides.test.js:
```diff
@@ -12,7 +12,7 @@ import {
   mergeOverrides,
   mergeOverride,
-  mergeStyleOverrides,
+  mergeSpecialOverrides,
```
can we rename the utility to something like 'mergeConfigurationOverrides'? the fields props and style are not particularly special (they are two-thirds of the available fields). the difference i see between these fields and 'component' is that these provide small configurations while the other fully replaces

chasestarr commented on src/helpers/__tests__/overrides.test.js:
```diff
@@ -35,6 +35,10 @@ describe('Helpers - Overrides', () => {
       style: {color: 'blue'},
     };
+    const overridePropsAsFunction = {
+      props: () => ({propName: 'propValue'}),
```
can we also get a test where it reads from its function arguments?
```js
{
  props: p => ({/* ...do something with p */}),
}
```

chasestarr commented on src/helpers/overrides.js:
```diff
@@ -14,9 +14,14 @@ export type StyleOverrideObjectT = {};
 
 export type StyleOverrideT = StyleOverrideObjectT | StyleOverrideFunctionT;
 
+export type PropsOverrideFunctionT = ({}) => ?{};
```
can we get a test asserting what happens when the return value of a props function is null/undefined?

chasestarr commented on src/helpers/overrides.js:
```diff
@@ -14,9 +14,14 @@ export type StyleOverrideObjectT = {};
+export type PropsOverrideFunctionT = ({}) => ?{};
+export type PropsOverrideObjectT = {};
+
+export type PropsOverrideT = PropsOverrideObjectT | PropsOverrideFunctionT;
```
This type (and its sub-types) exactly duplicate `StyleOverrideT`. Can it be de-duplicated into a `ConfigurationOverrideT` and alias `export type StyleOverrideT = ConfigurationOverrideT`?

qqingvarqq commented:
updated,
renamed `mergeStyleOverrides` to `mergeConfigurationOverrides`,
added more tests,
updated types, `PropsOverrideT` and `StyleOverrideT` replaced by `ConfigurationOverrideT` and alias  `StyleOverrideT`

chasestarr commented on src/helpers/__tests__/overrides.test.js:
```diff
@@ -138,20 +165,46 @@ describe('Helpers - Overrides', () => {
+    const override2 = {props: () => ({prop1: 'newValue'})};
+    const result = mergeOverride(override1, override2);
+    expect(typeof result.props).toBe('function');
+    // $FlowFixMe style should be a function here
```
what does this mean?

qqingvarqq replied:
Sorry, I forgot to change the comment. Since `mergeOverrideds` now can compose props which are functions I added test case for that. Similar to `mergeOverride can compose style functions`
And since props can be undefined flow warn us in this particular scenario when we call props.

jh3y commented:
Great stuff! This will close #1552 👍 
On a side note. There is also a proposal (#208) to consolidate the `overrides` utils into one function. I haven't fully dug into this yet, but wondering how much this PR steps into that territory.

qqingvarqq commented:
@jh3y This PR does not touch this proposal #208, it just makes it possible to pass a function into props overrides.
